tests/unit/src/Internal/LoggerTest.php

```diff
@@ -0,0 +1,300 @@
+	 *
+	 * @var WC_Payment_Gateway_WCPay|MockObject
+	 */
+	private $mock_gateway;
+
+
+	/**
+	 * Sets up the logger.
+	 */
+	protected function setUp(): void {
+		parent::setUp();
+
+		$this->mock_wc_logger = $this->createMock( WC_Logger::class );
+		$this->mode           = $this->createMock( Mode::class );
+		$this->mock_gateway   = $this->createMock( WC_Payment_Gateway_WCPay::class );
+		$this->logger         = $this->getMockBuilder( Logger::class )
+			->setConstructorArgs(
+				[
+					$this->mock_wc_logger,
+					$this->mode,
+					$this->mock_gateway,
+				]
+			)
+			->setMethods( [ 'can_log' ] )
+			->getMock();
+	}
+
+	/**
+	 * Test log method of the Logger class.
+	 */
+	public function test_log_success() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( true );
+		$this->mock_wc_logger
+			->expects( $this->once() )
+			->method( 'log' )
+			->with( WC_Log_Levels::INFO, 'Test Log 1' );
+		$this->logger->log( 'Test Log 1' );
+	}
+
+	/**
+	 * Test log method of the Logger class when can_log returns false.
+	 */
+	public function test_log_failure() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( false );
+		$this->mock_wc_logger
+			->expects( $this->never() )
+			->method( 'log' );
+		$this->logger->log( 'Test Log 1' );
+	}
+
+		/**
+		 * Test emergency method of the Logger class.
+		 */
+	public function test_log_level_emergency() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( true );
+		$this->mock_wc_logger
+			->expects( $this->once() )
+			->method( 'log' )
+			->with( WC_Log_Levels::EMERGENCY, 'Test Log Emergency' );
+		$this->logger->emergency( 'Test Log Emergency' );
+	}
+
+	/**
+	 * Test alert method of the Logger class.
+	 */
+	public function test_log_level_alert() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( true );
+		$this->mock_wc_logger
+			->expects( $this->once() )
+			->method( 'log' )
+			->with( WC_Log_Levels::ALERT, 'Test Log Alert' );
+		$this->logger->alert( 'Test Log Alert' );
+	}
+
+	/**
+	 * Test critical method of the Logger class.
+	 */
+	public function test_log_level_critical() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( true );
+		$this->mock_wc_logger
+			->expects( $this->once() )
+			->method( 'log' )
+			->with( WC_Log_Levels::CRITICAL, 'Test Log Critical' );
+		$this->logger->critical( 'Test Log Critical' );
+	}
+
+	/**
+	 * Test error method of the Logger class.
+	 */
+	public function test_log_level_error() {
+		$this->logger->expects( $this->once() )
+			->method( 'can_log' )
+			->willReturn( true );
+		$this->mock_wc_logger
+			->expects( $this->once() )
+			->method( 'log' )
+			->with( WC_Log_Levels::ERROR, 'Test Log Error' );
+		$this->logger->error( 'Test Log Error' );
+	}
+
+	/**
+	 * Test warning method of the Logger class.
+	 */
+	public function test_log_level_warning() {
```

Comment:
There are a lot of very similar tests here. Can we use `@dataProvider` instead?

Reply:
I considered it, but the method name is not the same that has to be tested. Am sure it can be done with some additional complexity passing the dynamic method name and calling that. But I thought it is better to avoid doing that and keep it simple. What do you think?